AsyncDisplayKit/Details/ASCollectionViewLayoutController.mm

```diff
@@ -164,4 +165,19 @@ - (CGRect)rangeBoundsWithScrollDirection:(ASScrollDirection)scrollDirection
   return CGRectExpandToRangeWithScrollableDirections(rect, tuningParameters, _scrollableDirections, scrollDirection);
 }
 
+- (ASRangeTuningParameters)tuningParametersForRangeType:(ASLayoutRangeType)rangeType
+{
+    BOOL isVisible = ASInterfaceStateIncludesVisible(_collectionView.interfaceState);
+    BOOL isScrolling = (_collectionView.scrollDirection != ASScrollDirectionNone);
+    // When the collecion view is not visible or not scrolling, make display range only as big as visible range.
+    // This reduce early creation of views and layers.
+    if (!isVisible || !isScrolling) {
+        if (rangeType == ASLayoutRangeTypeDisplay) {
+            return [super tuningParametersForRangeType:ASLayoutRangeTypeVisible];
```

Comment:
Technically, this behavior should be implemented in the range controller. It is fairly expensive to request the index paths that lie within a range, so if we are going to use an identical range, we should simply use the same NSSet from one request. The biggest question is how we could expand this range slightly with a minimum of cost. It probably makes sense to keep the independent calculation of the display range for now, whenever it is different than the visible range, even though that would be less efficient than simply reusing the visible range -- there is probably no generalizable way to extend the range "a bit" without running the full range index path calculation, at least for collection view custom layouts.  In the state where the view has never been scrolled before, defined the display range to be something like one quarter of the screen in each direction beyond visible, leading and trailing (of course scroll direction is unknown at this point, and in many cases the view may only be scrollable in one direction, so that will automatically ensure the other direction isn't adding cost).

A very important case to consider is the pager node. When we first open the page in the pager node, we ideally you would want that one page to load very quickly, while around the same time calling fetch data for the pages to the left and right, without including them in the display range. We need to create a new notification / delegate call of some kind to trigger when most of the loading for the visible area is done, and only at that time, extend the display range to include the left and right pages while it also extending the data range to include two pages to the left and right. In this state, the display range for the 1-left and 1-right pages should be identical to the visible area of those pages, allowing us to reduce calls to the collection layout to a maximum of 2.  We should also constrain the fetch data range to not go as far as it normally does for an item which isn't visible at all yet, e.g. right now for those left and right pages we may be fetching 2-3 screens vertically off the bottom even when the user may never begin viewing that page. It is only once those pages have become visible that the display range should include one quarter of the screen in each the leading and trailing direction, and once the user initiates the very first scroll on that view, that we should immediately expand them to the full / normal ranges.

This idea of scaling the ranges is fairly new, and how we handle ranges today does not have a great mechanism to do it. The right implementation may be to apply a percentage scale factor of some kind to the ranges, or to name these different kinds of states (maybe a bitfield encompassing states such as, offscreen by one, or on the screen but not yet scrolled in the direction of the range) to allow some degree of tuning in control over this behavior for different types of content that may have different requirements in terms of and resource consumption.  With a small number of new abstractions in this area, we should be able to automate certain optimization such as checking if a range is the same and only requesting from the layout controller that set of index paths once and using that set for any appropriate calculations in the range controller.